File: array/two_pointers/collision_pointers/0016_3sum_closest.js
/*
Given an integer array nums of length n and an integer target, find three 
integers in nums such that the sum is closest to target.

Return the sum of the three integers.

You may assume that each input would have exactly one solution.
*/

// https://leetcode.com/problems/3sum-closest/

// brute force
// time complexity: O(n^3)
// space complexity: O(1)
var threeSumClosest = function (nums, target) {
  let closest = Infinity;
  let diff = Infinity;
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      for (let k = j + 1; k < nums.length; k++) {
        let sum = nums[i] + nums[j] + nums[k];
        let newDiff = Math.abs(target - sum);
        if (newDiff < diff) {
          diff = newDiff;
          closest = sum;
        }
      }
    }
  }
  return closest;
};

// two pointers
// time complexity: O(n^2)
// space complexity: O(1)
var threeSumClosest = function (nums, target) {
  // sort the array
  nums.sort((a, b) => a - b);
  let closest = Infinity;
  let diff = Infinity;
  for (let i = 0; i < nums.length; i++) {
    // define two pointers
    let left = i + 1;
    let right = nums.length - 1;
    // end condition: left === right
    while (left < right) {
      let sum = nums[i] + nums[left] + nums[right];
      let newDiff = Math.abs(target - sum);
      if (newDiff < diff) {
        diff = newDiff;
        closest = sum;
      }
      // move the pointers
      if (sum < target) {
        left++;
      } else {
        right--;
      }
    }
  }
  return closest;
};

File: array/two_pointers/collision_pointers/0016_3sum_closest.py
# Given an integer array nums of length n and an integer target, find three integers in nums such that the sum is closest to target.

# Return the sum of the three integers.

# You may assume that each input would have exactly one solution.

# 3 <= nums.length <= 500
# -1000 <= nums[i] <= 1000
# -10**4 <= target <= 10**4

# https://leetcode.com/problems/3sum-closest/

# adding the duplicate skipping codes to this problem will result in a slowdown
# the cost of duplicate checkers > redundant calculations

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        l = len(nums)
        nums.sort()
        dis = float('inf')
        ans = 0
        for first in range(l - 2):
            # skip first duplicates
            # if first > 0 and nums[first] == nums[first - 1]: continue
            second, third = first + 1, l - 1
            while second < third:
                # skip second and third duplicates
                # while second < third and second > first + 1 and nums[second] == nums[second - 1]: second += 1
                # while third > second and third < l - 1 and nums[third] == nums[third + 1]: third -= 1
                # if second >= third: break
                total = nums[first] + nums[second] + nums[third]
                if abs(total - target) < dis:
                    dis = abs(total - target)
                    ans = total
                if total < target:
                    second += 1
                elif total > target:
                    third -= 1
                else:
                    # distance reach minimum (0) when total == target
                    return total  
                    # second += 1
                    # third -= 1
        return ans
    
